libsplinter/src/circuit/mod.rs

```diff
@@ -380,94 +381,167 @@ impl<'r> IntoIterator for &'r Roster {
     }
 }
 
+#[derive(Clone)]
 pub struct SplinterState {
     // location of the persisted state
     storage_location: String,
     // The state that is persisted
-    circuit_directory: CircuitDirectory,
+    circuit_directory: Arc<RwLock<CircuitDirectory>>,
     // Service id to Service that contains the node the service is connected to. Not persisted.
-    service_directory: HashMap<ServiceId, Service>,
+    service_directory: Arc<RwLock<HashMap<ServiceId, Service>>>,
 }
 
 impl SplinterState {
     pub fn new(storage_location: String, circuit_directory: CircuitDirectory) -> Self {
         SplinterState {
             storage_location,
-            circuit_directory,
-            service_directory: HashMap::new(),
+            circuit_directory: Arc::new(RwLock::new(circuit_directory)),
+            service_directory: Arc::new(RwLock::new(HashMap::new())),
         }
     }
 
     pub fn storage_location(&self) -> &str {
         &self.storage_location
     }
 
-    fn write_circuit_directory(&self) -> Result<(), WriteError> {
+    fn commit_circuit_directory(&self) -> Result<(), SplinterStateError> {
         // Replace stored state with the current splinter state
-        let mut storage = get_storage(self.storage_location(), || self.circuit_directory.clone())
-            .map_err(WriteError::GetStorageError)?;
+
+        let circuit_directory = self.circuit_directory()?;
+
+        let mut storage = get_storage(self.storage_location(), move || circuit_directory.clone())
+            .map_err(|err| SplinterStateError {
+            context: format!("Failed to get splinter storage: {}", err),
+            source: None,
+        })?;
 
         // when this is dropped the new state will be written to storage
-        **storage.write() = self.circuit_directory.clone();
+        **storage.write() = self.circuit_directory()?;
         Ok(())
     }
 
     // ---------- methods to access service directory ----------
-    pub fn service_directory(&self) -> &HashMap<ServiceId, Service> {
-        &self.service_directory
+
+    pub fn service_directory(&self) -> Result<HashMap<ServiceId, Service>, SplinterStateError> {
```

Comment:
This function might not make sense any more, if the API is expected to be sync.  The use case for this would be looking up a service by id, so perhaps this should be changed to `get_service(&self, service_id: ServiceId)`  This depends on usage.  This get all version should be used infrequently, as it requires a pretty large clone.